File: hashing/equal_row_and_column_pairs.py
from collections import defaultdict

def equalPairs(grid: List[List[int]]) -> int:
    def convert_to_key(arr: List[int]):
        return tuple(arr)
    
    row_dic = defaultdict(int)
    for row in grid:
        row_dic[convert_to_key(row)] += 1
    
    col_dic = defaultdict(int)
    for col in range(len(grid[0])):
        current_col = []
        for row in range(len(grid)):
            current_col.append(grid[row][col])
        
        col_dic[convert_to_key(current_col)] += 1
    
    ans = 0
    for arr in row_dic:
        ans += row_dic[arr] * col_dic[key]
    
    return ans

File: hashing/equal_row_and_column_pairs.java
public String convertToKey(int[] arr)
{
    StringBuilder sb = new StringBuilder();
    for (int i = 0; i < arr.length; i++)
    {
        sb.append(arr[i]);
        sb.append(",");
    }

    return sb.toString();
}

public int equalPairs(int[][] grid)
{
    Map<String, Integer> row_dic = new HashMap<>();
    for (int[] row : grid)
    {
        String key = convertToKey(row);
        row_dic.put(key, row_dic.getOrDefault(key, 0) + 1);
    }

    Map<String, Integer> col_dic = new HashMap<>();
    for (int col = 0; col < grid[0].length; col++)
    {
        int[] currentCol = new int[grid.length];
        for (int row = 0; row < grid.length; row++)
        {
            currentCol[row] = grid[row][col];
        }

        String key = convertToKey(currentCol);
        col_dic.put(key, col_dic.getOrDefault(key, 0) + 1);
    }

    int ans = 0;
    for (String key : row_dic.keySet())
    {
        ans += row_dic.get(key) * col_dic.getOrDefault(key, 0);
    }

    return ans;
}